cmd/vcfWebAnnotate/vcfWebAnnotate.go

```diff
@@ -0,0 +1,195 @@
+	"io"
+	"net/http"
+	"strings"
+)
+
+func usage() {
+	fmt.Print(
+		"vcfWebAnnotate - Annotate a vcf file by querying various databases via CellBase.\n\n" +
+			"Usage:\n" +
+			"  vcfWebAnnotate [options] in.vcf\n\n" +
+			"Options:\n\n")
+	flag.PrintDefaults()
+}
+
+func vcfWebAnnotate(data <-chan vcf.Vcf, header vcf.Header, outfile io.Writer, batchSize int, numBuffers int) {
+	header = addAnnotationHeader(header)
+	vcf.NewWriteHeader(outfile, header)
+	filledBufChan := make(chan []vcf.Vcf, numBuffers)
+	emptyBufChan := make(chan []vcf.Vcf, numBuffers)
+
+	for i := 0; i < numBuffers-1; i++ {
+		emptyBufChan <- make([]vcf.Vcf, 0, batchSize) // send all but 1 buffer to empty
+	}
+	buf := make([]vcf.Vcf, 0, batchSize)
+
+	go queryWorker(filledBufChan, emptyBufChan, outfile)
+
+	for v := range data { // read vcfs until you have a full batch then send for annotation
+		if len(buf) == batchSize {
+			filledBufChan <- buf
+			buf = <-emptyBufChan
+			buf = buf[:0]
+		}
+		buf = append(buf, v)
+	}
+
+	if len(buf) > 0 { // send any variants remaining in the buffer
+		filledBufChan <- buf
+	}
+	close(filledBufChan)
+
+	for range emptyBufChan { // stall until queryWorker is finished
+	}
+}
+
+func queryWorker(filledBufChan <-chan []vcf.Vcf, emptyBufChan chan<- []vcf.Vcf, outfile io.Writer) {
+	baseUrl := "http://bioinfo.hpc.cam.ac.uk/cellbase/webservices/rest/v4/hsapiens/genomic/variant/annotation?assembly=grch38"
+	query := new(bytes.Buffer)
+	var responses Responses
+	data := new(bytes.Buffer)
+
+	for buf := range filledBufChan { // get a slice of vcfs to query
+		query.Reset()
+		for i := range buf { // generates a comma separated list of variants in the url
+			if i > 0 {
+				query.WriteByte(',')
+			}
+			query.WriteString(fmt.Sprintf("%s:%d:%s:%s", buf[i].Chr, buf[i].Pos, buf[i].Ref, buf[i].Alt[0]))
+		}
+		response, err := http.Post(baseUrl, "text/plain", query) // query
+		exception.PanicOnErr(err)
+
+		data.Reset()
+		_, err = data.ReadFrom(response.Body)
+		exception.PanicOnErr(err)
+		err = json.Unmarshal(data.Bytes(), &responses)
+		exception.PanicOnErr(err)
+
+		annotateVcfs(buf, responses)
+		vcf.WriteVcfToFileHandle(outfile, buf)
+		emptyBufChan <- buf // return buffer for reuse
+	}
+	close(emptyBufChan)
+}
+
+func annotateVcfs(vcfs []vcf.Vcf, res Responses) {
+	var ann []string
+	var consequence Consequence
+	var proteinAnn ProteinAnnotation
+	var maxAf float64
+	for i := range vcfs {
+		ann = ann[:0] // reset
+
+		maxAf = getMaxPopAf(res.Responses[i])
+		if maxAf != -1 {
+			ann = append(ann, fmt.Sprintf("MaxPopAF=%.2g", maxAf))
+		}
+
+		if len(res.Responses[i].Results[0].Consequences) == 0 {
+			continue
+		}
+		consequence = res.Responses[i].Results[0].Consequences[0]
+
+		if res.Responses[i].Results[0].ConsequenceType != "" {
+			ann = append(ann, fmt.Sprintf("Consequence=%s", res.Responses[i].Results[0].ConsequenceType))
+		}
+
+		if consequence.GeneName != "" {
+			ann = append(ann, fmt.Sprintf("Gene=%s", consequence.GeneName))
```

Comment:
What happens if we overlap more than one gene/transcript/protein?

Reply:
Hmmm.... I think we should at least put a note in the usage statement that we do this.  I can image people wanting all vcfs that overlap a gene or something like that, and currently they will not get it (in corner cases).  Can we add an option for "allData" or something similar that will generate the huge field?  Does it still get unwieldy if you give all genes it overlaps and all protein annotations?  Then only the transcriptIds would be hidden by default?

Reply:
That sounds like a good solution for this pull request.  This might be a good discussion topic for our next code review so that you can fully explain the situation and we can see if anyone has a good idea.